guides/javascript.md

```diff
@@ -0,0 +1,176 @@
+# JavaScript
+
+**NOTE:** All of these guidelines apply to TypeScript as well.
+
+## Begin function/method names with verbs (except for `is`)
+
+When naming a function or method, describe the action, not the outcome.
+
+For example, instead of this:
+
+``` javascript
+function formattedChangelog() {
+  // ...
+}
+```
+
+say this:
+
+``` javascript
+function formatChangelog() {
+  // ...
+}
+```
+
+For a function or method that returns a boolean, prefix the name with a descriptive verb so that readers don't confuse it with a boolean variable. So instead of this:
+
+``` javascript
+class NetworkController {
+  isEIP1559Compatible() {
+    // ...
+  }
+}
+```
+
+say something like this:
+
+``` javascript
+class NetworkController {
+  determineEIP1559Compatibility() {
+    // ...
+  }
+}
+```
+
+### Sources/Inspiration
+
+> The name of a method is usually a verb or a verb phrase saying what the method *does*: `close`, `readPersons`. The name should also suggest if the method is mutating the object or returning a new one.
+>
+> _– [Kotlin coding conventions](https://kotlinlang.org/docs/coding-conventions.html#choose-good-names)_
+
+> Because methods are the means of taking action, the design guidelines require that method names be verbs or verb phrases.
+>
+> – _[.NET Fundamentals](https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/names-of-type-members#names-of-methods)_
+
+> Prefer method and function names that make [call] sites form grammatical English phrases.
+>
+> _– [Swift API Design Guidelines](https://www.swift.org/documentation/api-design-guidelines/#strive-for-fluent-usage)_
+
+> it's a best practice to actually tell _what the function is doing by giving the **function name a verb as prefix**_
+
+> _– [JavaScript Naming Convention Best Practices](https://javascript.plainenglish.io/javascript-naming-convention-best-practices-b2065694b7d)_
+
+## Begin boolean variable names with `has` or `is`
+
+When naming a variable that has a boolean type, prefix it with a form of "to be" or "to have" (e.g. `is*`, `has*`). If appropriate, include past or future tense and/or use an auxiliary verb such as `should`. If the name starts with `not`, invert the value. If the name already includes "to be" or "to have" but in the middle, move it to the beginning. Finally, take special note of variables which are created via the use of the React `useState` hook.
+
+For example, instead of this:
+
+``` javascript
+const [removed, setRemoved] = useState(false);
+const sendDisabled = standard !== ERC721;
+const notEnoughGas = !isBalanceSufficient(...);
+const currentTabIsConnectedToSelectedAddress = Boolean(
+  selectedAddressSubjectMap && selectedAddressSubjectMap[originOfCurrentTab]
+);
+```
+
+say this:
+
+``` javascript
+const [isRemoved, setIsRemoved] = useState(false);
+const isSendDisabled = standard !== ERC721;
+const hasEnoughGas = isBalanceSufficient(...);
```

Comment:
Not worth holding this up, but just wanna throw out something. For objects we'd like to enforce alphabetical assortment of keys, and sometimes that results in ordering that is not easily grepped for logic. Example:
```js
const state = {
  hasEnoughGas: false,
  isRecipientCopied: false,
  isGasSet: false,
};
```

in the above example, there is a unrelated state key injected between two quasi-related state keys.

```js
const state = {
  gasIsSet: false,
  gasCostIsCovered: false,
  recipientIsCopied: false,
};
```

The above example is off the top of my head, but the idea here is to put the noun/primary object first. 

Mainly commenting here to get your thoughts, not to suggest we change this. The written content already reflects an improvement in a lot of areas of our legacy code.

Reply:
Okay, I can get behind that. I know that you also proposed something similar here, is this related? https://github.com/MetaMask/metamask-extension/pull/14784#discussion_r882774247

Reply:
I just realized that my original comment had incorrect ordering lol and wouldn't have been a problem, but i'll leave it. This is why we need linting for this rule :P 

Yeah thats exactly what triggered this comment.

Reply:
Also if the noun first paradigm costs more overhead than the object values being not grouped together then we can avoid this convention entirely. Just one of those things that irks me that may be just a personal thing and not widely felt.

Reply:
I like the idea, but enforcing it would be challenging.

I'm reminded here of the Redux Style Guide, which has separate sections for "Essentials", "Strongly Recommended", and "Recommended".

Noun-first naming seems like a good candidate for the "Recommended" category.

I was considering creating a page on naming things; maybe I will try to fit this in that document.

Reply:
I like the idea of having those same categories for our best practices, and i'm fine with the noun first naming being "recommended"

Reply:
Okay, I've added a new guideline which incorporates your suggestion here: [f9caede](https://github.com/MetaMask/contributor-docs/pull/11/commits/f9caede572b83f19ec6da5570db83b611fc1d49b). However this includes a bunch of other changes, notably the addition of guideline categories, which I've extracted to this PR: https://github.com/MetaMask/contributor-docs/pull/12

Reply:
Update: I ended up marking this guideline with a 💡 to indicate that it's more of a suggestion with alternatives.